File: Frontend/src/types/tour.ts
export type TourStep = {
  title: string;
  explanation: string;
  file_path: string;
  start_line: number;
  end_line: number;
  snippet: string;
  why: string | null;
  language?: string | null;
};

export type TourFreshness = {
  indexed_sha: string | null;
  head_sha: string | null;
  commits_behind: number | null;
  measurable: boolean;
  changed_cited_files: string[];
  checked_at: string;
};

export type TourArtifact = {
  title: string;
  topic: string;
  repo_name: string;
  steps: TourStep[];
  freshness?: TourFreshness | null;
};

export type JourneyStatus =
  | "pending"
  | "generating"
  | "running"
  | "complete"
  | "failed"
  | "cancelled";

export type JourneyResponse = {
  id: number;
  status: JourneyStatus;
  repoName: string;
  ref: string | null;
  topic: string;
  artifact: TourArtifact | null;
  error: string | null;
};

export type JourneySummary = {
  id: number;
  status: JourneyStatus;
  repoName: string;
  ref: string | null;
  topic: string;
  createdAt: string;
};

File: Backend/app/models/tour.py
"""Structured tour artifact — the contract for guided-tour generation."""

from pydantic import BaseModel, Field, model_validator


class TourStep(BaseModel):
    title: str = Field(min_length=1)
    explanation: str = Field(min_length=1)
    file_path: str = Field(min_length=1)
    start_line: int = Field(ge=1)
    end_line: int = Field(ge=1)
    snippet: str = Field(min_length=1)
    why: str | None = None

    @model_validator(mode="after")
    def end_not_before_start(self) -> "TourStep":
        if self.end_line < self.start_line:
            raise ValueError("end_line must be >= start_line")
        return self


class TourArtifact(BaseModel):
    title: str = Field(min_length=1)
    topic: str = Field(min_length=1)
    repo_name: str = Field(min_length=1)
    steps: list[TourStep] = Field(min_length=1)
